src/quic/streams.cc

```diff
@@ -1,61 +1,1052 @@
+    if (ret < 0) {
+      MarkErrored();
+      std::move(next)(ret, nullptr, 0, [](int) {});
+      // Since we are erroring and won't be able to make use of this DataQueue
+      // any longer, let's free both it and the reader and put ourselves into
+      // an errored state. Further attempts to read from the outbound will
+      // result in a UV_EBADF error. The caller, however, should handle this by
+      // closing down the stream so that doesn't happen.
+      return ret;
+    }
+
+    if (ret == bob::Status::STATUS_EOS) {
+      // Here, we know we are done with the DataQueue and the Reader, but we
+      // might not yet have committed or acknowledged all of the queued data.
+      // We'll release our references to the queue_ and reader_ but everything
+      // else is untouched.
+      MarkEnded();
+      if (uncommitted_ > 0) {
+        // If the read returns eos, and there are uncommitted bytes in the
+        // queue, we'll set eos_ to true and return the current set of
+        // uncommitted bytes.
+        PullUncommitted(std::move(next));
+        return bob::STATUS_CONTINUE;
+      }
+      // If the read returns eos, and there are no uncommitted bytes in the
+      // queue, we'll return eos with no data.
+      std::move(next)(bob::Status::STATUS_EOS, nullptr, 0, [](int) {});
+      return bob::Status::STATUS_EOS;
+    }
+
+    if (ret == bob::Status::STATUS_BLOCK) {
+      // If the read returns blocked, and there are uncommitted bytes in the
+      // queue, we'll return the current set of uncommitted bytes.
+      if (uncommitted_ > 0) {
+        PullUncommitted(std::move(next));
+        return bob::Status::STATUS_CONTINUE;
+      }
+      // If the read returns blocked, and there are no uncommitted bytes in the
+      // queue, we'll return blocked.
+      std::move(next)(bob::Status::STATUS_BLOCK, nullptr, 0, [](int) {});
+      return bob::Status::STATUS_BLOCK;
+    }
+
+    // Reads here are generally expected to be synchronous. If we have a reader
+    // that insists on providing data asynchronously, then we'll have to block
+    // until the data is actually available.
+    if (ret == bob::Status::STATUS_WAIT) {
+      next_pending_ = true;
+      std::move(next)(bob::Status::STATUS_BLOCK, nullptr, 0, [](int) {});
+      return bob::Status::STATUS_BLOCK;
+    }
+
+    CHECK_EQ(ret, bob::Status::STATUS_CONTINUE);
+    PullUncommitted(std::move(next));
+    return bob::Status::STATUS_CONTINUE;
+  }
+
+  void MemoryInfo(MemoryTracker* tracker) const override {
+    tracker->TrackField("queue", queue_);
+    tracker->TrackField("reader", reader_);
+    tracker->TrackFieldWithSize("buffer", total_);
+  }
+
+  SET_MEMORY_INFO_NAME(Stream::Outbound)
+  SET_SELF_SIZE(Outbound)
+
+ private:
+  struct OnComplete {
+    bob::Done done;
+    explicit OnComplete(bob::Done done) : done(std::move(done)) {}
+    ~OnComplete() { std::move(done)(0); }
+  };
+
+  void PullUncommitted(bob::Next<ngtcp2_vec> next) {
```

Comment:
I think we should add some sort of context/comment to functions like this. It is getting a little bit hard to understand with no comments (or references).